src/Editor.ts

```diff
@@ -1,47 +1,89 @@
+ *
+ * @property filename The filename (from the current `project`) that the editor should be displaying for editing
+ * @property options Editor options that should be passed to the monaco editor when it is created
+ * @property onEditorInit Called when the monaco editor is created and initialized by the widget, passing the instance of the monaco editor
+ * @property onEditorLayout Called when the widget calls `.layout()` on the monaco editor
  */
-export default class Editor extends Destroyable {
-	private _currentFile: string;
-	private _editor: monaco.editor.IStandaloneCodeEditor;
+export interface EditorProperties extends WidgetProperties, ThemeableProperties {
+	filename?: string;
+	options?: monaco.editor.IEditorOptions;
+
+	onEditorInit?(editor: monaco.editor.IStandaloneCodeEditor): void;
+	onEditorLayout?(): void;
+}
 
+/* tslint:disable:variable-name */
+const EditorBase = ThemeableMixin(WidgetBase);
+/* tslint:enable:variable-name */
+
+@theme(css)
+export default class Editor extends EditorBase<EditorProperties> {
+	private _editor: monaco.editor.IStandaloneCodeEditor;
+	private _didChangeHandle: monaco.IDisposable;
 	private _onDidChangeModelContent = () => {
-		project.setFileDirty(this._currentFile);
+		if (this.properties.filename) {
+			project.setFileDirty(this.properties.filename);
+		}
 	}
+	private _queuedLayout = false;
 
-	/**
-	 * A class which is a simple abstraction of the `monaco-editor` editor
-	 * @param element The root HTML element to attach the editor to
-	 */
-	constructor (element: HTMLElement, options?: monaco.editor.IEditorOptions) {
-		super();
+	private _initEditor(element: HTMLDivElement) {
+		/* doing this async, during next macro task to help ensure the editor does a proper layout */
+		queueTask(() => {
+			this._editor = globalMonaco.editor.create(element, this.properties.options);
+			this._didChangeHandle = this._editor.onDidChangeModelContent(debounce(this._onDidChangeModelContent, 1000));
+			const { onEditorInit } = this.properties;
+			onEditorInit && onEditorInit(this._editor);
 
-		this._editor = globalMonaco.editor.create(element, options);
-		const didChangeHandle = this._editor.onDidChangeModelContent(debounce(this._onDidChangeModelContent, 1000));
+			this.own(createHandle(() => {
+				this._editor.dispose();
+				this._didChangeHandle.dispose();
+			}));
+		});
+	}
 
-		this.own(createHandle(() => {
-			this._editor.dispose();
-			didChangeHandle.dispose();
-		}));
+	public render() {
+		return v('div', {
+			afterCreate: this._initEditor,
+			afterUpdate: this.updateEditor as any,
+			classes: this.classes(css.base)
+		});
 	}
 
-	/**
-	 * Display a file in the editor from the currently loaded project
-	 * @param model The model to display
-	 */
-	display(filename: string): void {
-		if (!project.includes(filename)) {
-			throw new Error(`File "${filename}" is not part of the project.`);
+	@afterRender()
+	public updateEditor(node?: DNode): DNode | undefined {
```

Comment:
It was exactly the same code... I could move it to a totally private method and then call it from the two locations, but it seemed odd as well.